src/Core/NETPortable/NetPortableProfileTable.cs

```diff
@@ -14,95 +13,95 @@ public static class NetPortableProfileTable
-        // like it's done already for the _portableProfiles. Not sure why it's done that way there.
-        private static IDictionary<string, NetPortableProfile> _portableProfilesByCustomProfileString;
-        // key is the identifier of the optional framework and value is the list of tuple of (optional Framework Version, set of profiles in which they are optional)
-        private static IDictionary<string, List<VersionStringISetTuple>> _portableProfilesSetByOptionalFrameworks;
+        private static Lazy<CompiledNetPortableProfileCollection> _lazyPortableProfiles
+            = new Lazy<CompiledNetPortableProfileCollection>(() => new CompiledNetPortableProfileCollection(BuildPortableProfileCollection()));
+        private static CompiledNetPortableProfileCollection _portableProfiles;
+
+        private static CompiledNetPortableProfileCollection Compiled
+        {
+            get
+            {
+                return _portableProfiles ?? _lazyPortableProfiles.Value;
+            }
+        }
 
         public static NetPortableProfile GetProfile(string profileName)
         {
-            if (String.IsNullOrEmpty(profileName))
+            if (string.IsNullOrEmpty(profileName))
             {
                 throw new ArgumentException(CommonResources.Argument_Cannot_Be_Null_Or_Empty, "profileName");
             }
 
             // Original behavior fully preserved, as we first try the original behavior.
             // NOTE: this could be a single TryGetValue if this collection was kept as a dictionary...
-            if (Profiles.Contains(profileName))
+            if (Compiled.Profiles.Contains(profileName))
             {
-                return Profiles[profileName];
+                return Compiled.Profiles[profileName];
             }
 
             // If we didn't get a profile by the simple profile name, try now with 
             // the custom profile string (i.e. "net40-client")
             NetPortableProfile result = null;
-            _portableProfilesByCustomProfileString.TryGetValue(profileName, out result);
+            Compiled.PortableProfilesByCustomProfileString.TryGetValue(profileName, out result);
 
             return result;
         }
 
         internal static NetPortableProfileCollection Profiles
         {
-            get
+            set
             {
-                if (_portableProfiles == null)
+                if (value == null)
+                {
+                    _portableProfiles = null;
+                }
+                else
                 {
-                    // We use the setter so that we can consistently set both the 
-                    // existing collection as well as the CustomProfileString-indexed one.
-                    // This keeps both in sync.
-                    Profiles = BuildPortableProfileCollection();
-                    CreateOptionalFrameworksDictionary();
+                    _portableProfiles = new CompiledNetPortableProfileCollection(value);
```

Comment:
you can change it to 
```c#
_lazyPortableProfiles = new Lazy<CompiledNetPortableProfileCollection>(() => new CompiledNetPortableProfileCollection(value));
```
Then you can get rid of this `_portableProfiles` duality nonsense.